src/object/cli_obj.c

```diff
@@ -1509,6 +1521,415 @@ shard_task_reset_target(tse_task_t *shard_task, void *arg)
+						iods[i].iod_size;
+			unsigned long re = iods[i].iod_recxs[j].rx_nr *
+						iods[i].iod_size + rs - 1;
+
+			/* No partial-parity updates, so this function won't be
+			 * called if parity is present.
+			 */
+			D_ASSERT(!(PARITY_INDICATOR & rs));
+			for (ext_idx = rs; ext_idx <= re; ext_idx += len) {
+				unsigned int cell = ext_idx/len;
+
+				*tgt_set |= 1UL << (cell+p);
+				if (*tgt_set == full) {
+					*tgt_set = 0;
+					return;
+				}
+			}
+		}
+	}
+	/* In case it's a single value, set the tgt_set to send to all.
+	 * These go everywhere for now.
+	 */
+	if (*tgt_set == par_only)
+		*tgt_set = 0;
+}
+
+/* Iterates over the IODs in the update, encoding all full stripes contained
+ * within each recx.
+ */
+static int
+ec_obj_update(tse_task_t *task, daos_obj_id_t oid, daos_oclass_attr_t *oca,
+	      uint64_t *tgt_set)
+{
+	daos_obj_update_t	*args = dc_task_get_args(task);
+	struct ec_params	*head = NULL;
+	struct ec_params	*current = NULL;
+	unsigned int		 i, j;
+	int			 rc = 0;
+
+	for (i = 0; i < args->nr; i++) {
+		if (ec_has_full_stripe(&(args->iods[i]), oca, tgt_set)) {
+			struct ec_params *params;
+
+			D_ALLOC_PTR(params);
+			if (params == NULL) {
+				rc = -DER_NOMEM;
+				break;
+			}
+			ec_init_params(params, &args->iods[i], &args->sgls[i]);
+			if (head == NULL) {
+				head = params;
+				current = head;
+				rc = ec_set_head_params(head, args, i);
+				if (rc != 0)
+					break;
+			} else {
+				current->next = params;
+				current = params;
+			}
+			if (args->iods[i].iod_type == DAOS_IOD_ARRAY) {
+				unsigned int sg_idx = 0;
+				size_t sg_off = 0;
+
+				for (j = 0; j < args->iods[i].iod_nr; j++) {
+					rc = ec_array_encode(params, oid,
+							     &args->iods[i],
+							     &args->sgls[i],
+							     oca, j, &sg_idx,
+							     &sg_off);
+					if (rc != 0)
+						break;
+				}
+				rc = ec_update_params(params, &args->iods[i],
+						      &args->sgls[i],
+						      oca->u.ec.e_len);
+				head->iods[i] = params->niod;
+				head->sgls[i] = params->nsgl;
+				D_ASSERT(head->nr == i);
+				head->nr++;
+			} else {
+				D_ASSERT(args->iods[i].iod_type ==
+					 DAOS_IOD_SINGLE &&
+					 args->iods[i].iod_size >=
+					 oca->u.ec.e_len * oca->u.ec.e_k);
+				/* Encode single value */
+			}
+		} else if (head != NULL && &(head->sgls[i]) != NULL) {
+			/* Add sgls[i] and iods[i] to head. Since we're
+			 * adding ec parity (head != NULL) and thus need to
+			 * replace the arrays in the update struct.
+			 */
+			head->iods[i] = args->iods[i];
+			head->sgls[i] = args->sgls[i];
+			D_ASSERT(head->nr == i);
+			head->nr++;
+		}
+	}
+	if (rc != 0 && head != NULL) {
+		D_FREE(head->iods);
+		D_FREE(head->sgls);
+		while (head != NULL) {
+			current = head;
+			D_FREE(current->niod.iod_recxs);
+			D_FREE(current->nsgl.sg_iovs);
+			for (i = 0; i < current->p_segs.p_nr; i++)
+				D_FREE(current->p_segs.p_bufs[i]);
+			D_FREE(current->p_segs.p_bufs);
+			head = current->next;
+			D_FREE(current);
+		}
```

Comment:
seems a bit duplicate with ec_free_params_cb?

Reply:
The call back has to recover the head from the data parameter via a cast from void*, and set a return code based on the task result. Otherwise it's the same code.  I can create a function that is called from both places.